### PR TITLE
feat: APKAM enrollmentId support and atkeys_file refactor

### LINKED ISSUE #155: at_c: Breaking Changes Tracker
```[tasklist]
### Breaking Changes to make
- [ ] https://github.com/atsign-foundation/at_c/issues/154
- [ ] https://github.com/atsign-foundation/at_c/issues/140
- [ ] https://github.com/atsign-foundation/at_c/issues/225
- [ ] https://github.com/atsign-foundation/at_c/issues/127
- [ ] https://github.com/atsign-foundation/at_c/issues/128
- [ ] https://github.com/atsign-foundation/at_c/issues/129
- [ ] https://github.com/atsign-foundation/at_c/issues/284
- [ ] https://github.com/atsign-foundation/at_c/issues/375
- [ ] https://github.com/atsign-foundation/at_c/issues/377
```

### LINKED ISSUE #287: support auth with apkam atKeys file
### Is your feature request related to a problem? Please describe.

apkam atKeys files contain the APKAM enrollment ID which should be included in the pkam command that the client sends. The enrollment ID is required by the atServer in order to locate the APKAM public key for this enrollment, which it then uses to verify the signature.  Currently the enrollment ID is not sent, therefore the signature verification fails.

### Describe the solution you'd like

When reading atKeys json
- check if there is an enrollment ID in the json
- if an enrollment ID is present, then use it when generating the pkam command which is sent to the atServer

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### LINKED ISSUE #375: at_c: `atclient_monitor_start` should have `atclient_monitor_start_request_options`
We should have a `atclient_monitor_start_request_options` which has fields like
- char *regex
- long startTimestamp
- long endTimestamp

So the struct will look something like this:

```h
typedef struct atclient_monitor_start_request_options {
  char *regex;
  long startTimestamp;
  long endTimestamp;
} atclient_monitor_start_request_options;
```

I am unfamiliar with how the monitor verb and timstampts work, but I know this work 

This will be a breaking change as the `atclient_monitor_start` function signature will change to something like this:

```h
atclient_monitor_start(atclient *monitor_client, atclient_monitor_start_request_options *request_options);
```

### LINKED ISSUE #387: at_c: refactor `atclient_atkeysfile` to `atclient_atkeys_file`
To be in line with our conventions, we should do this refactor sooner than later.

### LINKED ISSUE #393: at_c: `message_type` and `message_body` from `monitor.c` `static int parse_message` function point to a freed address
The static `parse_message` function in `monitor.c` returns `message_type` and `message_body` pointers that point to memory that is freed at the end of the function. This function creates a copy of the original parameter using strdup, which internally uses malloc to allocate memory. However, the function then frees this allocated memory (`original_copy`) before returning, leaving `message_type` and `message_body` pointing to invalid memory.

https://github.com/atsign-foundation/at_c/blob/aaf943db9c9de6b28b64ebdf9903928401ef793a/packages/atclient/src/monitor.c#L247

The solution would involve returning to the old approach (use `original` parameter, and free it outside of the method after doing memcpy of the type and body), or removing the `free(original_copy)` from the exit of the method (which could lead to a memory leak if not handled properly later).

### PR DESCRIPTION
**Pull request by JeremyTubongbanua**
closes #287
closes #387
closes #393 
closes #375 
closes #155 

**- What I did**
- Added `enrollment_id` support to `atkeys_file.h/.c` and `atkeys.h/.c` 
- atclient now builds pkam verb with `enrollmentId:`, if it exists
- Refactored `atkeysfile` to `atkeys_file` to be consistent with conventions
- Refactored `atclient_pkam_authenticate_options` usage in `atclient_pkam_authenticate` and `atclient_monitor_pkam_authenticate`
  - Added functional test to ensure both `options = NULL` and predefined options work with pkam authentication

**- How to verify it**

The PKAM Authenticate example using keys that were APKAM'd

```bash
[INFO] 2024-09-08 00:15:39.234159 | pkam_authenticate | atclient_atkeys_file_read: 0
[INFO] 2024-09-08 00:15:39.235201 | pkam_authenticate | atclient_atkeys_populate_from_atkeys_file: 0
[INFO] 2024-09-08 00:15:39.235218 | atclient | options is NULL. Using production atDirectory to find atServer host and port
[DEBG] 2024-09-08 00:15:39.693942 | connection |        SENT: "smoothalligator"
[DEBG] 2024-09-08 00:15:39.797274 | connection |        RECV: "245b44d4-a4bd-5f33-b077-c559f956486a.swarm0001.atsign.zone:1722"
[INFO] 2024-09-08 00:15:39.797353 | atclient | atserver_host: 245b44d4-a4bd-5f33-b077-c559f956486a.swarm0001.atsign.zone
[INFO] 2024-09-08 00:15:39.797358 | atclient | atserver_port: 1722
[DEBG] 2024-09-08 00:15:40.457876 | connection |        SENT: "from:smoothalligator"
[DEBG] 2024-09-08 00:15:40.536135 | connection |        RECV: "data:_7ffed872-c543-43d9-8601-b134a6dbd6e2@smoothalligator:bfa908a0-14dd-4203-8879-28425ddd2fa9"
[DEBG] 2024-09-08 00:15:40.545910 | connection |        SENT: "pkam:enrollmentId:f4bdc0a1-2f76-479f-ae41-84275638ee95:XfWj9N2WbIqXBqA4vZxdSjnaBbDDddVCrAVZwWpsnv2kk9lwvlEUE7gDLzDPfqGYSrmwoswPDkQcs40t3Wp+Yh6HRFiUrefArB2wP5Q3l7rcc+A/hV9KL41JSBTkJXd9I93P1UkgM90y1l2WM/0/QIqkl14IvU0mqTlJw1ClyZ413qZG7iKBdYSqlzGOw1k4pfLQLLnDPW7GidnnSdC3A9Gtjwa74UzBYIZEipvApU+Hu9gnZ8PV0ygN4BCBJ/CCdGVwVh6V2Tq7a62bxEx6MKKhtdzGII9/NpNDiTAP42k3kkzdUr6CG2++ICQWag6AdLrQR7IWgKtiCQoBkutFng=="
[DEBG] 2024-09-08 00:15:40.634723 | connection |        RECV: "data:success"
[DEBG] 2024-09-08 00:15:40.634739 | pkam_authenticate | Authenticated
```

**- Description for the changelog**
feat: APKAM enrollment support

### CODE REVIEW

**Comment by JeremyTubongbanua** [comment]
@gkc @srieteja Can I get your confirmation this is how the PKAM command with an enrollmentId looks like?

```
[DEBG] 2024-09-08 00:15:40.545910 | connection |        SENT: "pkam:enrollmentId:f4bdc0a1-2f76-479f-ae41-84275638ee95:XfWj9N2WbIqXBqA4vZxdSjnaBbDDddVCrAVZwWpsnv2kk9lwvlEUE7gDLzDPfqGYSrmwoswPDkQcs40t3Wp+Yh6HRFiUrefArB2wP5Q3l7rcc+A/hV9KL41JSBTkJXd9I93P1UkgM90y1l2WM/0/QIqkl14IvU0mqTlJw1ClyZ413qZG7iKBdYSqlzGOw1k4pfLQLLnDPW7GidnnSdC3A9Gtjwa74UzBYIZEipvApU+Hu9gnZ8PV0ygN4BCBJ/CCdGVwVh6V2Tq7a62bxEx6MKKhtdzGII9/NpNDiTAP42k3kkzdUr6CG2++ICQWag6AdLrQR7IWgKtiCQoBkutFng=="
```

**Comment by gkc** [comment]
@sitaram-kalluri @murali-shris see above question from @JeremyTubongbanua

**Comment by sitaram-kalluri** [comment]
@JeremyTubongbanua : Yes, It is correct. Also, PKAM command optionally accepts signing algo and hashing algo
> @gkc @srieteja Can I get your confirmation this is how the PKAM command with an enrollmentId looks like?
> 
> ```
> [DEBG] 2024-09-08 00:15:40.545910 | connection |        SENT: "pkam:enrollmentId:f4bdc0a1-2f76-479f-ae41-84275638ee95:XfWj9N2WbIqXBqA4vZxdSjnaBbDDddVCrAVZwWpsnv2kk9lwvlEUE7gDLzDPfqGYSrmwoswPDkQcs40t3Wp+Yh6HRFiUrefArB2wP5Q3l7rcc+A/hV9KL41JSBTkJXd9I93P1UkgM90y1l2WM/0/QIqkl14IvU0mqTlJw1ClyZ413qZG7iKBdYSqlzGOw1k4pfLQLLnDPW7GidnnSdC3A9Gtjwa74UzBYIZEipvApU+Hu9gnZ8PV0ygN4BCBJ/CCdGVwVh6V2Tq7a62bxEx6MKKhtdzGII9/NpNDiTAP42k3kkzdUr6CG2++ICQWag6AdLrQR7IWgKtiCQoBkutFng=="
> ```

**Review by gkc — COMMENTED** [review]
@JeremyTubongbanua there are changes in atclient.c regarding host and port which seem to be unrelated to the main purpose of this PR ... can you add explanation of those changes to the PR description please?

**Review by gkc — COMMENTED** [review]
@JeremyTubongbanua also monitor's message parsing has changed, can you add description of those changes also please?

**Comment by xavierchanth** [comment]
I'm picking up this PR, will add the appropriate notes to the PR body & work on getting this mergable

File: packages/atclient/src/atclient.c [hunk 1]
```diff
@@ -369,6 +383,7 @@ exit: {
   free(root_cmd);
   free(from_cmd);
   free(pkam_cmd);
+  free(atserver_host);
```
Comment (xavierchanth) [inline packages/atclient/src/atclient.c]: Note to self: potential bug, for me to verify later.

Comment (xavierchanth) [reply packages/atclient/src/atclient.c]: fixed in [d78b8fd](https://github.com/atsign-foundation/at_c/pull/400/commits/d78b8fd277257c096d2479fbdd8785dc430301d3)

File: examples/desktop/at_talk/src/main.c [hunk 2]
```diff
@@ -235,7 +235,7 @@ static void *monitor_handler(void *xargs) {
   const char *from_atsign = args->from_atsign;
 
   pthread_mutex_lock(&monitor_mutex);
-  if ((ret = atclient_monitor_pkam_authenticate(monitor, atserver_host, atserver_port, atkeys, from_atsign)) != 0) {
+  if ((ret = atclient_monitor_pkam_authenticate(monitor, from_atsign, atkeys, NULL)) != 0) {
```
Comment (xavierchanth) [inline examples/desktop/at_talk/src/main.c:238]: This signature is changed, host and port are now part of the options struct (which can be NULL, and the pkam_authenticate will look it up for you)

File: examples/desktop/at_talk/src/main.c [hunk 3]
```diff
@@ -4,7 +4,7 @@
 #include <atchops/iv.h>
 #include <atclient/atclient.h>
 #include <atclient/atclient_utils.h>
-#include <atclient/atkeysfile.h>
+#include <atclient/atkeys_file.h>
```
Comment (xavierchanth) [inline examples/desktop/at_talk/src/main.c:7]: This header file was renamed to conform to our naming standards

File: examples/desktop/repl/src/main.c
Comment (xavierchanth) [inline examples/desktop/repl/src/main.c:1]: Naming standards changes

File: packages/atclient/CMakeLists.txt [hunk 4]
```diff
@@ -150,9 +150,8 @@ if(NOT ESP_PLATFORM)
   )
 
   # EXPORT
-  if(NOT ATCLIENT_AS_SUBPROJECT)
+  if(NOT ATCLIENT_AS_SUBPROJECT AND ATCLIENT_EXPORT)
     # Export the library
-    export(PACKAGE ${PROJECT_NAME})
```
Comment (xavierchanth) [inline packages/atclient/CMakeLists.txt:155]: Don't export the package by default

File: packages/atclient/include/atclient/atkeys.h [hunk 5]
```diff
@@ -13,12 +13,14 @@
...
 #define ATCLIENT_ATKEYS_ENCRYPT_PRIVATE_KEY_INDEX 0
 #define ATCLIENT_ATKEYS_SELF_ENCRYPTION_KEY_INDEX 0
+#define ATCLIENT_ATKEYS_ENROLLMENT_ID_INDEX 0
 
 #define ATCLIENT_ATKEYS_PKAM_PUBLIC_KEY_INITIALIZED (VALUE_INITIALIZED << 0)
 #define ATCLIENT_ATKEYS_PKAM_PRIVATE_KEY_INITIALIZED (VALUE_INITIALIZED << 1)
 #define ATCLIENT_ATKEYS_ENCRYPT_PUBLIC_KEY_INITIALIZED (VALUE_INITIALIZED << 2)
 #define ATCLIENT_ATKEYS_ENCRYPT_PRIVATE_KEY_INITIALIZED (VALUE_INITIALIZED << 3)
 #define ATCLIENT_ATKEYS_SELF_ENCRYPTION_KEY_INITIALIZED (VALUE_INITIALIZED << 4)
+#define ATCLIENT_ATKEYS_ENROLLMENT_ID_INITIALIZED (VALUE_INITIALIZED << 5)
```
Comment (xavierchanth) [inline packages/atclient/include/atclient/atkeys.h:23]: Constants to define the position & value of enrollment id in our is_initialized boolean vector

File: packages/atclient/include/atclient/atkeys.h [hunk 6]
```diff
@@ -32,20 +34,22 @@
...
+  char *encrypt_public_key_base64;               // base64 encoded, RSA-2048 key, decrypted
   atchops_rsa_key_public_key encrypt_public_key; // contains n, e
 
-  char *encrypt_private_key_base64;               // base64 encoded, RSA-2048 key, decrypted
+  char *encrypt_private_key_base64;                // base64 encoded, RSA-2048 key, decrypted
   atchops_rsa_key_private_key encrypt_private_key; // conatins n, e, d, p, q
 
   char *self_encryption_key_base64; // base64 encoded, AES-256 key, decrypted
 
+  char *enrollment_id;
```
Comment (xavierchanth) [inline packages/atclient/include/atclient/atkeys.h:51]: Add enrollment id to the keys structure

File: packages/atclient/include/atclient/atkeys.h [hunk 7]
```diff
@@ -64,46 +68,51 @@ void atclient_atkeys_init(atclient_atkeys *atkeys);
...
 
 int atclient_atkeys_set_encrypt_private_key_base64(atclient_atkeys *atkeys, const char *encrypt_private_key_base64,
-                                                const size_t encryptprivatekeybase64len);
+                                                   const size_t encryptprivatekeybase64len);
 
 int atclient_atkeys_set_self_encryption_key_base64(atclient_atkeys *atkeys, const char *selfencryptionkeybase64,
-                                                const size_t selfencryptionkeybase64len);
+                                                   const size_t selfencryptionkeybase64len);
+
+int atclient_atkeys_set_enrollment_id(atclient_atkeys *atkeys, const char *enrollment_id, const size_t enrollment_id_len);
```
Comment (xavierchanth) [inline packages/atclient/include/atclient/atkeys.h:85]: Enrollment id setter

File: packages/atclient/include/atclient/atkeys.h [hunk 8]
```diff
@@ -64,46 +68,51 @@ void atclient_atkeys_init(atclient_atkeys *atkeys);
...
 int atclient_atkeys_populate_encrypt_private_key(atclient_atkeys *atkeys, const char *encrypt_private_key_base64,
-                                               const size_t encryptprivatekeybase64len);
+                                                 const size_t encryptprivatekeybase64len);
 
 bool atclient_atkeys_is_pkam_public_key_base64_initialized(atclient_atkeys *atkeys);
 bool atclient_atkeys_is_pkam_private_key_base64_initialized(atclient_atkeys *atkeys);
 bool atclient_atkeys_is_encrypt_public_key_base64_initialized(atclient_atkeys *atkeys);
 bool atclient_atkeys_is_encrypt_private_key_base64_initialized(atclient_atkeys *atkeys);
 bool atclient_atkeys_is_self_encryption_key_base64_initialized(atclient_atkeys *atkeys);
+bool atclient_atkeys_is_enrollment_id_initialized(atclient_atkeys *atkeys);
```
Comment (xavierchanth) [inline packages/atclient/include/atclient/atkeys.h:104]: enrollment id is initialized getter

File: packages/atclient/include/atclient/atkeys.h [hunk 9]
```diff
@@ -118,28 +127,39 @@ bool atclient_atkeys_is_self_encryption_key_base64_initialized(atclient_atkeys *
 int atclient_atkeys_populate_from_strings(atclient_atkeys *atkeys, const char *aes_pkam_public_key_str,
                                           const size_t aes_pkam_public_key_len, const char *aes_pkam_private_key_str,
                                           const size_t aes_pkam_private_key_len, const char *aes_encrypt_public_key_str,
-                                          const size_t aes_encrypt_public_key_len, const char *aes_encrypt_private_key_str,
+                                          const size_t aes_encrypt_public_key_len,
+                                          const char *aes_encrypt_private_key_str,
                                           const size_t aes_encrypt_private_key_len, const char *self_encryption_key_str,
-                                          const size_t self_encryption_key_len);
+                                          const size_t self_encryption_key_len, const char *enrollment_id_str, const size_t enrollment_id_str_len);
```
Comment (xavierchanth) [inline packages/atclient/include/atclient/atkeys.h:133]: add enrollment id to the populate from strings method

File: packages/atclient/include/atclient/atkeys_file.h
Comment (xavierchanth) [inline packages/atclient/include/atclient/atkeys_file.h:1]: This file was renamed (but was marked as delete / add in git) from atkeysfile.h to atkeys_file.h, it also has identical changes as noted in atkeys.h

File: packages/atclient/include/atclient/monitor.h [hunk 10]
```diff
@@ -95,16 +94,13 @@ void atclient_monitor_free(atclient *monitor_conn);
...
- * @param atkeys the atkeys to use for the pkam authentication
  * @param atsign the atsign to use for the pkam authentication
+ * @param atkeys the atkeys to use for the pkam authentication
+ * @param options the options to use for the pkam authentication
  * @return int 0 on success
  */
-int atclient_monitor_pkam_authenticate(atclient *monitor_conn, const char *atserver_host, const int atserver_port,
-                                       const atclient_atkeys *atkeys, const char *atsign);
+int atclient_monitor_pkam_authenticate(atclient *monitor_conn, const char *atsign, const atclient_atkeys *atkeys,
+                                       atclient_pkam_authenticate_options *options);
```
Comment (xavierchanth) [inline packages/atclient/include/atclient/monitor.h:103]: Same introduction of options to monitor specific pkam authenticate implementation

File: packages/atclient/src/atclient.c [hunk 11]
```diff
@@ -191,10 +189,6 @@ int atclient_pkam_authenticate(atclient *ctx, const char *atsign, const atclient
                                atclient_pkam_authenticate_options *options) {
 
   int ret = 1; // error by default
-  if(options == NULL){
-    atclient_pkam_authenticate_options options;
-    atclient_pkam_authenticate_options_init(&options);
-  }
```
Comment (xavierchanth) [inline packages/atclient/src/atclient.c:197]: options is allowed to be NULL now

File: packages/atclient/src/atclient.c [hunk 12]
```diff
@@ -204,16 +198,6 @@ int atclient_pkam_authenticate(atclient *ctx, const char *atsign, const atclient
...
-  /*
-   * 1.1 If atserver_host and atserver_port are null and 0 respectively, then fetch the server host and port.
-   */
-  if (options->atserver_host == NULL && options->atserver_port == 0) {
-    if (atclient_utils_find_atserver_address(options->at_directory_host, options->at_directory_port, atsign,
-                                             &(options->atserver_host), &(options->atserver_port)) != 0) {
-      return ret;
-    }
-  }
-
```
Comment (xavierchanth) [inline packages/atclient/src/atclient.c:216]: This scenario is readdressed below

File: packages/atclient/src/atclient.c [hunk 13]
```diff
@@ -254,24 +241,44 @@ int atclient_pkam_authenticate(atclient *ctx, const char *atsign, const atclient
...
+   * 4. Get atserver_host and atserver_port
+   */
+  bool should_free_atserver_host = false;
+  if (options != NULL) {
+    if (atclient_pkam_authenticate_options_is_atdirectory_host_initialized(options) &&
+        atclient_pkam_authenticate_options_is_atdirectory_port_initialized(options)) {
+      atserver_host = options->atdirectory_host;
+      atserver_port = options->atdirectory_port;
+    }
+  }
```
Comment (xavierchanth) [inline packages/atclient/src/atclient.c:254]: Try to populate host and port if provided.

File: packages/atclient/src/atclient.c [hunk 14]
```diff
@@ -254,24 +241,44 @@ int atclient_pkam_authenticate(atclient *ctx, const char *atsign, const atclient
...
+                 "Missing atServer host or port. Using production atDirectory to look up atServer host and port\n");
+    if ((ret = atclient_utils_find_atserver_address(ATCLIENT_ATDIRECTORY_PRODUCTION_HOST,
+                                                    ATCLIENT_ATDIRECTORY_PRODUCTION_PORT, atsign, &atserver_host,
+                                                    &atserver_port)) != 0) {
+      atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "atclient_utils_find_atserver_address: %d\n", ret);
+      goto exit;
+    }
+    // only free this memory if it was allocated internally (by atclient_utils_find_atserver_address)
+    should_free_atserver_host = true;
   }
```
Comment (xavierchanth) [inline packages/atclient/src/atclient.c:267]: If host and port weren't populated from options, look them up.

File: packages/atclient/src/atclient.c [hunk 15]
```diff
@@ -321,17 +328,28 @@ int atclient_pkam_authenticate(atclient *ctx, const char *atsign, const atclient
...
-  snprintf(pkam_cmd, pkam_cmd_size, "pkam:%s\r\n", signature_base64);
+  size_t pos = 0;
+  pos += snprintf(pkam_cmd + pos, pkam_cmd_size - pos, "pkam:");
+
+  if (atclient_atkeys_is_enrollment_id_initialized((atclient_atkeys *)atkeys) && atkeys->enrollment_id != NULL) {
+    pos += snprintf(pkam_cmd + pos, pkam_cmd_size - pos, "enrollmentId:%s:", atkeys->enrollment_id);
+  }
+
+  pos += snprintf(pkam_cmd + pos, pkam_cmd_size - pos, "%s\r\n", signature_base64);
```
Comment (xavierchanth) [inline packages/atclient/src/atclient.c:349]: modify pkam verb to conditionally include the enrollment id

File: packages/atclient/src/atclient.c [hunk 16]
```diff
@@ -369,6 +387,9 @@ exit: {
   free(root_cmd);
   free(from_cmd);
   free(pkam_cmd);
+  if (should_free_atserver_host) {
+    free(atserver_host);
+  }
```
Comment (xavierchanth) [inline packages/atclient/src/atclient.c:392]: free the atserver_host if it was allocated during automatic lookup (wasn't provided by options)

File: packages/atclient/src/atkey.c
Comment (xavierchanth) [inline packages/atclient/src/atkey.c:1]: formatting and compiler warnings

File: packages/atclient/src/atkeys.c
Comment (xavierchanth) [inline packages/atclient/src/atkeys.c:1]: added enrollment id to atkeys struct and appropriate accessors/mutators (just following the same patterns we have been)

File: packages/atclient/src/atkeys.c [hunk 17]
```diff
@@ -622,57 +672,90 @@ int atclient_atkeys_populate_from_strings(atclient_atkeys *atkeys, const char *a
...
   }
 
+  // 6. enrollment id, if it exists
+  if (enrollment_id_str != NULL && enrollment_id_str_len > 0) {
+    if ((ret = atclient_atkeys_set_enrollment_id(atkeys, enrollment_id_str, enrollment_id_str_len)) != 0) {
+      atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "atclient_atkeys_set_enrollment_id: %d\n", ret);
+      goto exit;
+    }
+  }
+
```
Comment (xavierchanth) [inline packages/atclient/src/atkeys.c:712]: populate from strings if enrollment_id_str isn't NULL or empty

File: packages/atclient/src/atkeys.c [hunk 18]
```diff
@@ -622,57 +672,90 @@ int atclient_atkeys_populate_from_strings(atclient_atkeys *atkeys, const char *a
...
+             atkeys, atkeys_file->aes_pkam_public_key_str, strlen(atkeys_file->aes_pkam_public_key_str),
+             atkeys_file->aes_pkam_private_key_str, strlen(atkeys_file->aes_pkam_private_key_str),
+             atkeys_file->aes_encrypt_public_key_str, strlen(atkeys_file->aes_encrypt_public_key_str),
+             atkeys_file->aes_encrypt_private_key_str, strlen(atkeys_file->aes_encrypt_private_key_str),
+             atkeys_file->self_encryption_key_str, strlen(atkeys_file->self_encryption_key_str),
+             atkeys_file->enrollment_id_str, strlen(atkeys_file->enrollment_id_str))) != 0) {
+      atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR,
+                   "atclient_atkeys_populate_from_strings: %d | failed to populate from strings\n", ret);
+      goto exit;
+    }
```
Comment (xavierchanth) [inline packages/atclient/src/atkeys.c:745]: conditionally add the enrollment id if it is initialized when populating from atkeys file

File: packages/atclient/src/atkeys.c [hunk 19]
```diff
@@ -681,48 +764,81 @@ exit: { return ret; }
...
+  }
+
+  goto exit;
+exit: {
+  atclient_atkeysfile_free(&atkeys_file);
+  return ret;
+}
+}
+
+int atclient_atkeys_populate_from_string(atclient_atkeys *atkeys, const char *file_string) {
```
Comment (xavierchanth) [inline packages/atclient/src/atkeys.c:789]: new populate variant, similar to populate from path, but is canonical to the atclient_atkeys_file_from_string variant for atkeys_file

File: packages/atclient/src/atkeys.c [hunk 20]
```diff
@@ -798,15 +929,23 @@ static void unset_encrypt_private_key_base64(atclient_atkeys *atkeys) {
...
 }
 
+static void unset_enrollment_id(atclient_atkeys *atkeys) {
+  if (is_enrollment_id_initialized(atkeys)) {
+    free(atkeys->enrollment_id);
+  }
+  atkeys->enrollment_id = NULL;
+  set_enrollment_id_initialized(atkeys, false);
+}
+
```
Comment (xavierchanth) [inline packages/atclient/src/atkeys.c:946]: unset enrollment id, wiping value and removing from initialization vector

File: packages/atclient/src/atkeys.c [hunk 21]
```diff
@@ -915,28 +1054,59 @@ exit: { return ret; }
...
 
   set_self_encryption_key_base64_initialized(atkeys, true);
 
   ret = 0;
-  goto exit;
-exit: { return ret; }
+  return ret;
+}
+
+static int set_enrollment_id(atclient_atkeys *atkeys, const char *enrollment_id, const size_t enrollment_id_len) {
```
Comment (xavierchanth) [inline packages/atclient/src/atkeys.c:1085]: enrollment id setter

File: packages/atclient/src/atkeys_file.c
Comment (xavierchanth) [inline packages/atclient/src/atkeys_file.c:1]: renamed from atkeysfile.c

File: packages/atclient/src/atkeys_file.c [hunk 22]
```diff
@@ -0,0 +1,664 @@
...
+  }
+
+  ret = atclient_atkeysfile_from_string(atkeys_file, (const char *)readbuf);
+  return ret;
+}
+
+int atclient_atkeysfile_from_string(atclient_atkeysfile *atkeys_file, const char *file_string) {
+  int ret = 1;
+
+  cJSON *root = cJSON_Parse(file_string);
```
Comment (xavierchanth) [inline packages/atclient/src/atkeys_file.c:76]: atclient_atkeysfile_from_string variant: we cJSON_Parse the file at file_string, then do all the same, using cjson to extract the atkeys file contents.

File: packages/atclient/src/atnotification.c
Comment (xavierchanth) [inline packages/atclient/src/atnotification.c:1]: formatting and compiler warnings

File: packages/atclient/src/connection.c
Comment (xavierchanth) [inline packages/atclient/src/connection.c:1]: formatting and compiler warnings

File: packages/atclient/src/monitor.c [hunk 23]
```diff
@@ -40,14 +38,11 @@ void atclient_monitor_response_free(atclient_monitor_response *message) {
 void atclient_monitor_init(atclient *monitor_conn) { atclient_init(monitor_conn); }
 void atclient_monitor_free(atclient *monitor_conn) { atclient_free(monitor_conn); }
 
-int atclient_monitor_pkam_authenticate(atclient *monitor_conn, const char *atserver_host, const int atserver_port,
-                                       const atclient_atkeys *atkeys, const char *atsign) {
+int atclient_monitor_pkam_authenticate(atclient *monitor_conn, const char *atsign, const atclient_atkeys *atkeys,
+                                       atclient_pkam_authenticate_options *options) {
```
Comment (xavierchanth) [inline packages/atclient/src/monitor.c:42]: same signature change as mentioned before

File: packages/atclient/src/monitor.c [hunk 24]
```diff
@@ -40,14 +38,11 @@ void atclient_monitor_response_free(atclient_monitor_response *message) {
...
   int ret = 1;
 
-  atclient_pkam_authenticate_options options;
-  atclient_pkam_authenticate_options_init(&options);
-  ret = atclient_pkam_authenticate(monitor_conn, atsign, atkeys, &options);
-  if (ret != 0) {
+  if ((ret = atclient_pkam_authenticate(monitor_conn, atsign, atkeys, options)) != 0) {
```
Comment (xavierchanth) [inline packages/atclient/src/monitor.c:45]: defer to the other pkam_authenticate method - monitor's variant is just a wrapper around the normal pkam auth

File: packages/atclient/src/monitor.c [hunk 25]
```diff
@@ -150,10 +145,8 @@ int atclient_monitor_read(atclient *monitor_conn, atclient *atclient, atclient_m
     goto exit;
   }
 
-  int i = 0;
-  while (buffer[i] != ':') {
-    i++;
-  }
```
Comment (xavierchanth) [inline packages/atclient/src/monitor.c:156]: this was a hack to align the monitor messages, which should be resolved by the parse_message changes below

File: tests/functional_tests/tests/test_atclient_pkam_authenticate.c [hunk 26]
```diff
@@ -0,0 +1,118 @@
...
+    return ret;
+  }
+
+  ret += test1_pkam_no_options();
+  ret += test2_pkam_with_options();
+
+  return ret;
+}
+
+static int test1_pkam_no_options() {
```
Comment (xavierchanth) [inline tests/functional_tests/tests/test_atclient_pkam_authenticate.c:45]: Test pkam when options = NULL

File: tests/functional_tests/tests/test_atclient_pkam_authenticate.c [hunk 27]
```diff
@@ -0,0 +1,118 @@
...
+    atlogger_log(tag, ATLOGGER_LOGGING_LEVEL_ERROR, "Failed to authenticate\n");
+    return ret;
+  } else {
+    atlogger_log(tag, ATLOGGER_LOGGING_LEVEL_DEBUG, "Authenticated\n");
+  }
+
+  return ret;
+}
+
+static int test2_pkam_with_options() {
```
Comment (xavierchanth) [inline tests/functional_tests/tests/test_atclient_pkam_authenticate.c:78]: Test pkam auth when we've cached atServer connection info in `options`

**Comment by xavierchanth** [comment]
Tested here using C sshnpd: https://asciinema.org/a/n5MWtuUxF2lAGD0oSk7E0Wsc3

**Review by gkc — APPROVED** [review]
LGTM

**Pull request merged**